node_modules/col-activities/lib/notifications/daily.js

```diff
@@ -0,0 +1,548 @@
+/* Activity aggregation */
+
+/**
+ * Get the relevant activities for a user
+ *
+ * @param  {Object}       data    The relevant course data
+ * @param  {User}         user    The user for which to get the activities
+ * @return {Activity[]}           A set of activities for which the user can be notified
+ * @api private
+ */
+var getActivitiesForUser = function(data, user) {
+  // This variable will contain all activities that can be emailed to the user
+  var emailActivities = [];
+
+  // Asset comments
+  _.each(data.assets, function(asset) {
+    // Whether the user owns the asset
+    var isOwner = asset.users[user.id];
+
+    // All top comments made by other users
+    var topOtherComments = _.filter(asset.comments, function(comment) {
+      return (!comment.parent && comment.user_id !== user.id);
+    });
+
+    // All replies made to the current user
+    var replies = _.filter(asset.comments, function(comment) {
+      return (comment.parent && comment.parent.user_id === user.id);
+    });
+
+    // All replies made to the current user by other users
+    var otherReplies = _.filter(replies, function(reply) {
+      return (reply.user_id !== user.id);
+    });
+
+    // If we own the asset, we want all the new comments (that weren't made by ourselves)
+    if (isOwner && !_.isEmpty(topOtherComments)) {
+      emailActivities.push(new AssetActivity(asset, 'asset_comment'));
+    }
+
+    // If there's a reply on a comment of ours we only want those replies
+    else if (!_.isEmpty(replies) && !_.isEmpty(otherReplies)) {
+      emailActivities.push(new AssetActivity(asset, 'asset_comment_reply', replies));
+    }
+  });
+
+  // Add whiteboards that have new chat messages and the user is a member of
+  _.each(data.whiteboards, function(whiteboard) {
+    var otherChatMessages = _.filter(whiteboard.chats, function(chat) {
+      return (chat.user_id !== user.id);
+    });
+
+    if (_.find(whiteboard.users, {'id': user.id}) && !_.isEmpty(otherChatMessages)) {
+      emailActivities.push(new WhiteboardActivity(whiteboard))
+    }
+  });
+
+  return _.chain(emailActivities)
+    .sortByOrder(['lastActivity'], ['desc'])
+    .value();
+
+};
+
+/**
+ * An Asset activity
+ *
+ * @param {Asset}       asset       The asset on which there's some activity
+ * @param {String}      type        One of `asset_comment` or `asset_comment_reply`
+ * @param {Comment[]}   [replies]   In case it's an `asset_comment_reply` activity, this should be the set of comments that are replies to a comment of the current user
+ */
+var AssetActivity = function(asset, type, replies) {
+  // Take a copy of the asset
+  asset = _.extend({}, asset);
+
+  if (type === 'asset_comment_reply') {
+    asset.comments = replies;
+  }
+
+  // Retain the number of comments or replies that were made
+  var newComments = asset.comments.length;
+
+  // Flatten the comments
+  var comments = replies || asset.comments;
+  var commentsById = {};
+  _.each(comments, function(comment) {
```

Comment:
It's a way of copying the comments as the next step will start modifying them. If any comment is re-used across (users their) activities, it would mess up the level